### ISSUE TITLE
Discord Bot implementation?

### ISSUE THREAD

**Issue by shadowfox87** [SugoiFactory/SkillKeeper#6]
This is a great program. One of a kind. The only program on the net that uses challonge api and then calculates Trueskill ratings. What if you could automate this process in a Discord Bot? The challonge api lets you start a tournament, add a participant to a tournament, attach a file to a match, and update match results. These could all be done by a bot. The bot can also message each participant of who their next opponent is.

**Comment by themickeynick** [comment]
Why would you want to manage a tournament through discord? It seems very arbitrary and not really a great use of anyone's time. While we wait for challonge's API to update, it's better to work through what we can at this moment, plus it's kind of already stable (though I have some issues with the interface). Challonge is already easy to add players by providing a link in your discord. Automating the process is also time consuming to get running 100% properly as you have to know when to poll for updates or starting a downstream job after completing a match (which you can't really do unless you had access to Challonge's backend).
 
TL;DR - It's really not time efficient nor really worth anything as Challonge is pretty easy to use already and this application is already really easy to use.

PS:
One thing I would like to see or implement myself would be to get a web uploading system where you can upload the resulting files to some page (I mean CSV is an alright format, but it's annoying to parse). IDK, it might be better to just write an interface between challonge and a python web application using the TrueSkill package in this particular scenario.

**Comment by shadowfox87** [comment]
You can use the challonge API to attach a file to each match already. The reason it is useful is because the tournament host has to coordinate the Discord users and the challonge users. Hence, if a participant uses a bot to register within the Discord chat, then the bot has already linked the challonger username and the Discord username.

In any case, if you don't think it's time efficient, can you add the WHR ranking system which has been shown to be superior to Trueskill? A Ruby implementation can be found here: https://github.com/goshrine/whole_history_rating

**Comment by themickeynick** [comment]
But is it really worth anything to link Discord and Challonge usernames? What's the use case? Wouldn't the users themselves have to setup that link with the bot before being able to enter on Challonge? I think it is up to the TO to make those calls on making sure players know when to play and manage the tournament, leaving stuff like that to a bot sometimes ruins the interactions and the leniency of a human judge.

This almost seems like a project in and of itself to conjoin the three applications with SkillKeeper being a side effect of the discord/Challonge connection. I would suggest that there be a login system/API key storage for the application as well as on startup/login checking for new brackets not in the world.

WHR is okay, but not as tested as TrueSkill. TrueSkill has more support and does an okay job. In the paper itself, the author claims that there should be more research into the validity vs TTT and Edo.

**Comment by embell** [comment]
I agree, it doesn't sound like this would be directly related to SkillKeeper. Other ranking algorithms are probably viable as options if they need the same kind of data, but would probably not be trivial to incorporate.

**Comment by shadowfox87** [comment]
Well, I guess, I thought there might be a way to automate this process since it seems so repetitive. In cases when there is a knowledgeable TO and that TO has time, he/she can do all of this. However, in cases where a user wants to start a tournament and may not know how, this would be an easy for them. The users don't have to setup a link with the bot. They simply have to be already registered on challonge. The bot is the one that generates the challonge link and adds the participant from Discord to the tournament. Normally, a user has to go to the sign-up page on challonge.com to do this, but this way the bot is doing it. It makes life for a TO easier or for a person who is not a expert TO. The results of that tournament can then be converted to Trueskill rankings and updated automatically given a http link or as you said, a csv output.

**Comment by themickeynick** [comment]
Yeah SkillKeeper seems to be a side effect of the idea you want. You want a bot that links Discord and Challonge together with the resulting matches getting sent to SkillKeeper. While there is a library that SugoiFactory developed that's in this project, you still need to write the connections for everything else and is a larger undertaking than I think you think it is.

**Comment by shadowfox87** [comment]
Oh well, it's just an idea. I understand that I may not fully understand what you guys would have to go through to implement such a feature. I'm still using SkillKeeper manually as a TO, but I'm not always around. Sometimes, there is no TO and people want to continue a ranking system for their game and update the ladder.

**Comment by eaaliprantis** [comment]
It's actually worth it to include for tournaments.  I back up shadowfox87.  I am apart of multiple communities that need tournament bots (i created one but was trashed due to copyright/issues with a partner of mine).  So combining both of these together would be great.

**Comment by eaaliprantis** [comment]
Nick, i have ideas as well to make this all happen.  I have worked on a queue system for people checking in, etc.  And pushing/updating challonge, etc.  This can work out.  Contact me at eaaliprantis@gmail.com
